### PR TITLE
IOBuffer back to Ruby

### PR DESCRIPTION
**Pull request by nateberkopec**
File under "weird idea".

### CODE REVIEW

File: lib/puma/io_buffer.rb [hunk 1]
```diff
@@ -1,4 +1,19 @@
...
+    def initialize
+      if RUBY_VERSION < "2.2"
+        super
+      else
+        super(capacity: 4096)
+      end
+    end
+
+    def append(*args)
+      args.each { |a| concat(a) }
```
Comment (eregon) [inline lib/puma/io_buffer.rb]: FWIW, Ruby 2.4+ supports multi-args `String#concat`.

**Comment by eregon** [comment]
This looks great to me:
-155 lines of C
-72 lines of Java
+17 lines of Ruby

I guess the main concern is: how does it affect performance?

File: lib/puma/io_buffer.rb (hunk 1 above)
Comment (nateberkopec) [reply lib/puma/io_buffer.rb]: Nice! We have to support 2.2+ for now.

**Comment by nateberkopec** [comment]
@eregon Yeah, agree. I've got an open issue on #1981 surrounding that.

**Comment by kares** [comment]
a bit ugly that `IOBuffer` now has a "full" `String` API but I guess its not a user API anyway ...

**Comment by nateberkopec** [comment]
This branch, extra large request body:
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.96ms    1.02ms  16.99ms   72.38%
    Req/Sec     1.04k    98.31     1.22k    79.08%
  Latency Distribution
     50%    1.90ms
     75%    2.52ms
     90%    3.24ms
     99%    4.86ms
  124562 requests in 1.00m, 127.62GB read
Requests/sec:   2074.35
Transfer/sec:      2.13GB
```

Master, large request body:
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.74ms    0.86ms  16.93ms   71.69%
    Req/Sec     1.17k    66.38     1.29k    85.25%
  Latency Distribution
     50%    1.68ms
     75%    2.26ms
     90%    2.81ms
     99%    3.99ms
  139603 requests in 1.00m, 143.03GB read
Requests/sec:   2325.64
Transfer/sec:      2.38GB
```

This branch, hello world:
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   242.78us  137.14us   2.63ms   97.29%
    Req/Sec     8.67k   661.24     9.04k    96.52%
  Latency Distribution
     50%  224.00us
     75%  258.00us
     90%  290.00us
     99%    1.06ms
  173345 requests in 10.11s, 12.56MB read
Requests/sec:  17154.11
Transfer/sec:      1.24MB
```

Master, hello world:

```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   235.08us  135.54us   3.09ms   97.58%
    Req/Sec     8.90k     0.91k    9.36k    98.02%
  Latency Distribution
     50%  217.00us
     75%  251.00us
     90%  283.00us
     99%    1.03ms
  178961 requests in 10.10s, 12.97MB read
Requests/sec:  17715.35
Transfer/sec:      1.28MB
```

Maybe a 5% slowdown but that's really within the statistical margin of error. I think this is good for a merge.

**Comment by nateberkopec** [comment]
This appears to just stall the ruby 2.3 build 😖

File: lib/puma/io_buffer.rb [hunk 2]
```diff
@@ -1,4 +1,19 @@
 # frozen_string_literal: true
 
-require 'puma/detect'
-require 'puma/puma_http11'
+module Puma
+  class IOBuffer < String
+    def initialize
+      if RUBY_VERSION <= "2.3"
```
Comment (eregon) [inline lib/puma/io_buffer.rb]: This condition is incorrect (e.g. `2.3.4` <= `2.3` doesn't hold), `set capacity if RUBY_VERSION >= 2.4` should work fine instead, and that's likely what makes the 2.3 build fails:
https://github.com/puma/puma/pull/1980/checks?check_run_id=258129791
```
/home/runner/work/puma/puma/lib/puma/io_buffer.rb:9:in `initialize': unknown keyword: capacity (ArgumentError)
	from /home/runner/work/puma/puma/lib/puma/io_buffer.rb:9:in `initialize'
	from /home/runner/work/puma/puma/lib/puma/thread_pool.rb:97:in `new'
	from /home/runner/work/puma/puma/lib/puma/thread_pool.rb:97:in `block (2 levels) in spawn_thread'
	from /home/runner/work/puma/puma/lib/puma/thread_pool.rb:97:in `map'
	from /home/runner/work/puma/puma/lib/puma/thread_pool.rb:97:in `block in spawn_thread'
```

**Comment by nateberkopec** [comment]
I'm not sure the capacity optimization is worth it. Removed it and it's now benchmarking faster than master, and hello.sh benchmark on this branch is now `Requests/sec:  18001.34` vs master `Requests/sec:  16311.53`. Big response becnhmark is `Requests/sec:   2269.64` vs master's `Requests/sec:   2191.85`. So we are now even or even slightly faster than master branch 😆

**Comment by headius** [comment]
@nateberkopec Where are the benchmarks you ran above? I'd like to confirm there's no performance hit on JRuby and see if any tweaks might help improve things further.

**Comment by headius** [comment]
This is failing on JRuby because the HTTP parser ext still references the Java IOBuffer. I'll have a patch shortly.

**Comment by headius** [comment]
Just remove import:

```diff
diff --git a/ext/puma_http11/PumaHttp11Service.java b/ext/puma_http11/PumaHttp11Service.java
index da7c2719..00f63aa2 100644
--- a/ext/puma_http11/PumaHttp11Service.java
+++ b/ext/puma_http11/PumaHttp11Service.java
@@ -6,7 +6,6 @@ import org.jruby.Ruby;
 import org.jruby.runtime.load.BasicLibraryService;
 
 import org.jruby.puma.Http11;
-import org.jruby.puma.IOBuffer;
 import org.jruby.puma.MiniSSL;
 
 public class PumaHttp11Service implements BasicLibraryService {
```

**Comment by dentarg** [comment]
@headius maybe these? https://github.com/puma/puma/tree/v4.3.0/benchmarks/wrk

**Comment by headius** [comment]
On a simple Roda benchmark, this change appears to have no negative performance impact for JRuby. It's within the margin of error for this wrk-driven Roda benchmark. 👍

**Pull request merged**